osrf/vrx
# Port the perception scoring plugin

Let's wait until #523 is merged so that we can use some of the bridge infrastructure there.

### How to test it?

1. Launch the simulation loading the new `perception_task` world:
```
ros2 launch vrx_gz competition.launch.py world:=perception_task
```

2. Verify that at `t=25`, two buoys (red and green) are spawned ~8m in front of the WAM-V. The buoys should stay there for another 5 seconds, and then, they should disappear from the field of view.

3. Verify that at `t=35`, one buoy (black) is spawned ~8m in front of the WAM-V. The buoy should stay there for another 5 seconds, and then, it should disappear from the field of view.

4. The simulation should stop at `t=40`.

![perception_task](https://user-images.githubusercontent.com/1440739/193855882-9c2c8b93-fb79-4021-a262-538a353d0c8d.gif)

For testing the perception reports:

1. Create a directory for the client:
```
mkdir ~/perception_client & cd perception_client
```
2. Download these three files from [here](https://gist.github.com/caguero/2dd7ae577ab4d535e0c4d9f762e44163).
3. Create a build directory and compile:
```
mkdir build && cd build
cmake ..
make
```
4. Launch the simulation as the previous (1) step.
5. When you see the first round of objects, launch the client to report the red buoy:
```
. /opt/ros/humble/setup.bash
./perception_client
```
6. Wait until the simulation ends and look for the perception summary:
```
[ruby $(which gz) sim-1] [Dbg] [PerceptionScoringPlugin.cc:486] PerceptionScoring: 
[ruby $(which gz) sim-1] 	Name: red_0
[ruby $(which gz) sim-1] 	Type: mb_marker_buoy_red
[ruby $(which gz) sim-1] 	Time: 5
[ruby $(which gz) sim-1] 	Duration: 5
[ruby $(which gz) sim-1] 	Orig pose: -200 100 1 0 0 0
[ruby $(which gz) sim-1] 	Trial pose: 8 0 1 0 0 0
[ruby $(which gz) sim-1] 	Entity: 112
[ruby $(which gz) sim-1] 	Active: false
[ruby $(which gz) sim-1] 	Error: 0.0057877
[ruby $(which gz) sim-1] 
[ruby $(which gz) sim-1] [Dbg] [PerceptionScoringPlugin.cc:486] PerceptionScoring: 
[ruby $(which gz) sim-1] 	Name: green_0
[ruby $(which gz) sim-1] 	Type: mb_marker_buoy_green
[ruby $(which gz) sim-1] 	Time: 5
[ruby $(which gz) sim-1] 	Duration: 5
[ruby $(which gz) sim-1] 	Orig pose: -200 102 1 0 0 0
[ruby $(which gz) sim-1] 	Trial pose: 8 2 1 0 0 0
[ruby $(which gz) sim-1] 	Entity: 117
[ruby $(which gz) sim-1] 	Active: false
[ruby $(which gz) sim-1] 	Error: 10
[ruby $(which gz) sim-1] 
[ruby $(which gz) sim-1] [Dbg] [PerceptionScoringPlugin.cc:486] PerceptionScoring: 
[ruby $(which gz) sim-1] 	Name: black_0
[ruby $(which gz) sim-1] 	Type: mb_marker_buoy_black
[ruby $(which gz) sim-1] 	Time: 15
[ruby $(which gz) sim-1] 	Duration: 5
[ruby $(which gz) sim-1] 	Orig pose: -200 104 1 0 0 0
[ruby $(which gz) sim-1] 	Trial pose: 8 -2 1 0 0 0
[ruby $(which gz) sim-1] 	Entity: 122
[ruby $(which gz) sim-1] 	Active: false
[ruby $(which gz) sim-1] 	Error: 10
[ruby $(which gz) sim-1] 
[ruby $(which gz) sim-1] [Dbg] [PerceptionScoringPlugin.cc:490] Perception run score: 6.6686
```

Notice how the `mb_markler_buoy_red` buoy has an almost `0` error and the overall score is `6.66`. If you don't send the report you should get a `10`.

`Note`: It should be possible and way easier to send the report via `ros2 topic pub` but my autocomplete didn't really help me. I had a lot of escape characters and I couldn't fill the message properly.

~~**Note:** I'll update these instructions when #523 is merged to report the objects via ROS 2.~~
